File: src/test/java/com/example/demo/service/TimesheetServiceTest.java
package com.example.demo.service;

import com.example.demo.mapper.TimesheetEntryMapper;
import com.example.demo.model.TimesheetEntry;
import com.example.demo.model.TimesheetSaveCommand;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.context.ApplicationEventPublisher;

import java.time.LocalDate;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class TimesheetServiceTest {

    @Mock
    private TimesheetEntryMapper timesheetEntryMapper;

    @Mock
    private ApplicationEventPublisher eventPublisher;

    @InjectMocks
    private TimesheetService service;

    @Test
    void saveOrUpdateWithFlagsClearsIrregularFieldsAndPersistsEmptyNote() {
        LocalDate date = LocalDate.of(2026, 8, 30);
        TimesheetEntry existing = TimesheetEntry.builder()
                .id(10L)
                .userName("alice")
                .workDate(date)
                .note("午前休")
                .irregularWorkType("在宅")
                .irregularWorkDesc("旧値")
                .irregularWorkData("旧データ")
                .build();
        when(timesheetEntryMapper.selectByUserAndDate("alice", date)).thenReturn(existing);
        when(timesheetEntryMapper.updateTimes(any())).thenReturn(1);

        TimesheetSaveCommand command = TimesheetSaveCommand.builder()
                .userName("alice")
                .workDate(date)
                .noteProvided(true)
                .note("")
                .clearIrregular(true)
                .build();

        TimesheetEntry result = service.saveOrUpdateWithFlags(command);

        assertEquals("", result.getNote());
        assertEquals(null, result.getIrregularWorkType());
        assertEquals(null, result.getIrregularWorkDesc());
        assertEquals(null, result.getIrregularWorkData());
        verify(timesheetEntryMapper).updateTimes(any(TimesheetEntry.class));
        verify(eventPublisher).publishEvent(any(TimesheetUpdatedEvent.class));
    }
}

File: src/test/kotlin/com/example/demo/service/TimesheetServiceTest.kt
package com.example.demo.service

import com.example.demo.mapper.TimesheetEntryMapper
import com.example.demo.model.TimesheetEntry
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import java.time.LocalDate
import java.time.LocalTime

class TimesheetServiceTest {
    private val mapper = Mockito.mock(TimesheetEntryMapper::class.java)
    private val service = TimesheetService(mapper)

    @Test
    fun calcNormalDay() {
        val entry = TimesheetEntry(
            workDate = LocalDate.now(),
            userName = "u",
            startTime = LocalTime.of(9, 0),
            endTime = LocalTime.of(18, 0),
            breakMinutes = 60
        )
        val recalc = service.saveOrUpdate("u", entry.workDate, entry.startTime, entry.endTime, entry.breakMinutes)
        assertEquals(540, recalc.durationMinutes) // 9h
        assertEquals(480, recalc.workingMinutes) // 8h
    }

    @Test
    fun calcOverMidnight() {
        val today = LocalDate.now()
        val start = LocalTime.of(22, 30)
        val end = LocalTime.of(1, 15) // 翌日
        val recalc = service.saveOrUpdate("u", today, start, end, 30)
        assertEquals(165, recalc.durationMinutes) // 2h45m = 165
        assertEquals(135, recalc.workingMinutes)
    }

    @Test
    fun calcNegativeBreakClamp() {
        val today = LocalDate.now()
        val start = LocalTime.of(10, 0)
        val end = LocalTime.of(11, 0)
        val recalc = service.saveOrUpdate("u", today, start, end, -20)
        assertEquals(60, recalc.durationMinutes)
        assertEquals(60, recalc.workingMinutes) // break 0 に矯正
    }

    @Test
    fun calcTooLongSpanIgnored() {
        val today = LocalDate.now()
        val start = LocalTime.of(8, 0)
        val end = LocalTime.of(7, 30) // 24h を超えるケース(23h30 は許容、ここは -30 wrap=1430 正常) -> 正常
        val recalc = service.saveOrUpdate("u", today, start, end, 0)
        assertEquals(1430, recalc.durationMinutes)
    }
}

